presto-native-execution/presto_cpp/main/types/FunctionMetadata.cpp

```diff
@@ -0,0 +1,253 @@
+    // types, and as a vector function for complex types like DECIMAL. So do not
+    // throw an error if function metadata is not found in simple function
+    // signature map.
+    return simpleFunctionMetadata.back().first;
+  }
+
+  auto vectorFunctionMetadata = exec::getVectorFunctionMetadata(name);
+  if (vectorFunctionMetadata.has_value()) {
+    return vectorFunctionMetadata.value();
+  }
+  return std::nullopt;
+}
+
+const protocol::RoutineCharacteristics getRoutineCharacteristics(
+    const FunctionSignature& signature,
+    const std::string& name,
+    const protocol::FunctionKind& kind) {
+  protocol::Determinism determinism;
+  protocol::NullCallClause nullCallClause;
+  if (kind == protocol::FunctionKind::SCALAR) {
+    auto metadata = getScalarMetadata(name);
+    VELOX_USER_CHECK(
+        metadata.has_value(), "Metadata for function {} not found", name);
+    determinism = metadata.value().deterministic
+        ? protocol::Determinism::DETERMINISTIC
+        : protocol::Determinism::NOT_DETERMINISTIC;
+    nullCallClause = metadata.value().defaultNullBehavior
+        ? protocol::NullCallClause::RETURNS_NULL_ON_NULL_INPUT
+        : protocol::NullCallClause::CALLED_ON_NULL_INPUT;
+  } else {
+    // Default metadata values of DETERMINISTIC and CALLED_ON_NULL_INPUT for
+    // non-scalar functions.
+    determinism = protocol::Determinism::DETERMINISTIC;
+    nullCallClause = protocol::NullCallClause::CALLED_ON_NULL_INPUT;
+  }
+
+  protocol::RoutineCharacteristics routineCharacteristics;
+  routineCharacteristics.language =
+      std::make_shared<protocol::Language>(protocol::Language({"CPP"}));
+  routineCharacteristics.determinism =
+      std::make_shared<protocol::Determinism>(determinism);
+  routineCharacteristics.nullCallClause =
+      std::make_shared<protocol::NullCallClause>(nullCallClause);
+  return routineCharacteristics;
+}
+
+void updateFunctionMetadata(
+    const std::string& name,
+    const std::string& schema,
+    const FunctionSignature& signature,
+    protocol::JsonBasedUdfFunctionMetadata& metadata) {
+  metadata.docString = name;
+  metadata.schema = schema;
+  metadata.outputType = signature.returnType().toString();
+
+  const std::vector<TypeSignature> types = signature.argumentTypes();
+  std::vector<std::string> paramTypes;
+  for (const auto& type : types) {
+    paramTypes.emplace_back(type.toString());
+  }
+  metadata.paramTypes = paramTypes;
+}
+
+const std::vector<protocol::JsonBasedUdfFunctionMetadata> buildScalarMetadata(
+    const std::string& name,
+    const std::string& schema,
+    const std::vector<const FunctionSignature*>& signatures) {
+  std::vector<protocol::JsonBasedUdfFunctionMetadata> metadataList;
+  metadataList.reserve(signatures.size());
+  const protocol::FunctionKind kind = protocol::FunctionKind::SCALAR;
+
+  for (const auto& signature : signatures) {
+    protocol::JsonBasedUdfFunctionMetadata metadata;
```

Comment:
This loop code is repeated for scalar, aggregates and window functions. It can be abstracted into a common function reused across all 3. Add a parameter to control to add aggregation metadata for aggregates.